dropwizard-jersey/src/test/java/io/dropwizard/jersey/jackson/JsonProcessingExceptionMapperTest.java

```diff
@@ -97,7 +97,10 @@ public void returnsA400ForNonDeserializableRequestEntities() throws Exception {
 
     @Test
     public void returnsA400ForWrongInputType() throws Exception {
```

Comment:
Still scratching my head, so I've decided to do a brain dump. I think the root cause is that there is not a great way to differentiate between various errors of JsonMappingException

Examples from the JsonProcessingExceptionMapper test cases:

## Client Errors :

- Expecting an object, but JSON is `false` throws a JsonMappingException of "Can not construct instance of foo: no boolean/Boolean-argument constructor/factory method to deserialize from boolean value (false)". This should be a client error because the wrong json type was submitted.
- Expecting a URI, but JSON is `"no-scheme.com"`, throws a JsonMappingException of "Unexpected IOException (of type java.net.MalformedURLException): no protocol: no-scheme.com". Ideally there would be a way to reach the underlying cause, but `JsonMappingException` is the root. Other exceptions while deserializing can be uncovered (like `DateTimeException`). Looks to only affect test case `returnsA400ForMalformedInputCausingIoException`
- Expecting a list but given a string throws a JsonMappingException of "Can not deserialize instance of java.util.ArrayList out of VALUE_STRING token"
- Deserializing a DateTime from `[1,1,1,1]` throws a JsonMappingException of "Unexpected token (VALUE_NUMBER_INT), expected END_ARRAY: Expected array to end."

## Server Errors:

- `BrokenRepresentation` is missing a constructor and the exception is a JsonMappingException "Can not construct instance of BrokenRepresentation: no suitable constructor found, can not deserialize from Object value (missing default constructor or creator, or perhaps need to add/enable type information?)"
- `NonBeanImplementation` doesn't contain any information on serialization and exception is a JsonMappingException "No serializer found for class NonBeanImplementation and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS)"
- Trying to construct from an interface throws a JsonMappingException of "Can not construct instance of IInterface: abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information"

These seven situations all use `JsonMappingException` as the root cause and this is where we had to introduce heuristics. The following code gets us pretty close, but doesn't work on the client errors mentioned above.

```java
final JsonMappingException ex = (JsonMappingException) exception;
final Throwable cause = Throwables.getRootCause(ex);

// Exceptions that denote an error on the client side
final boolean clientCause = cause instanceof InvalidFormatException ||
    cause instanceof PropertyBindingException ||
    cause.getClass() != JsonMappingException.class;

if (!clientCause) {
    LOGGER.error("Unable to serialize or deserialize the specific type", exception);
    return Response.serverError().build();
}
```